cpp/src/fil/infer.cu

```diff
@@ -177,92 +212,173 @@ struct tree_aggregator_t {
+// allreduce_shmem ensures no race conditions
+template <typename Iterator, typename BinaryOp>
+__device__ __forceinline__ auto allreduce_shmem(Iterator begin, Iterator end,
+                                                BinaryOp op,
+                                                void* tmp_storage) {
+  typedef typename std::iterator_traits<Iterator>::value_type value_type;
+  value_type thread_partial;
+  for (Iterator it = begin + threadIdx.x; it < end; it += blockDim.x)
+    thread_partial = op(thread_partial, *it);
+  __syncthreads();  // free shared memory begin..end
+  auto res = block_reduce(thread_partial, op, tmp_storage);
+  // broadcast sum to all threads
+  __syncthreads();  // free up tmp_storage
+  if (threadIdx.x == 0) *(value_type*)tmp_storage = res;
+  __syncthreads();
+  return *(value_type*)tmp_storage;
+}
 
-  __device__ __forceinline__ finalize_block(void* tmp_storage_,
-                                            int num_classes_)
-    : tmp_storage(tmp_storage_), num_classes(num_classes_) {}
+// tmp_storage may overlap shared memory addressed by begin..end
+template <typename Iterator>
+__device__ __forceinline__ void write_best_class(Iterator begin, Iterator end,
+                                                 void* tmp_storage, float* out,
+                                                 int num_rows) {
+  // reduce per-class candidate margins to one best class candidate
+  // per thread (for each of the NITEMS rows)
+  auto best = to_vec(0, *begin);
+  for (int c = threadIdx.x; c < end - begin; c += blockDim.x)
+    best = vectorized(ArgMax())(best, to_vec(c, begin[c]));
+  __syncthreads();  // begin..end may overlap tmp_storage
+  // find best class per block (for each of the NITEMS rows)
+  best = block_reduce(best, vectorized(ArgMax()), tmp_storage);
+  // write it out to global memory
+  if (threadIdx.x > 0) return;
+#pragma unroll
+  for (int row = 0; row < num_rows; ++row) out[row] = best[row].key;
+}
 
-  template <int NITEMS>
-  static size_t smem_footprint() {
-    return sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
-  }
+/// needed for softmax
+__device__ float shifted_exp(float margin, float max) {
+  return expf(margin - max);
+}
 
-  template <int NITEMS>
-  __device__ __forceinline__ void write_best_class_in_block(
-    vec<NITEMS, best_margin_label> best, int valid_threads, float* out,
-    int num_rows) {
-    // find best class per block (for each of the NITEMS rows)
-    typedef BlockReduceBestClass<NITEMS> BlockReduceT;
-    best = BlockReduceT(*(typename BlockReduceT::TempStorage*)tmp_storage)
-             .Reduce(best, ArgMax(), valid_threads);
-    // write it out to global memory
-    if (threadIdx.x > 0) return;
+// TODO: fix tmp_storage placement
+// tmp_storage may NOT overlap shared memory addressed by begin..end
+template <typename Iterator>
+__device__ __forceinline__ void block_softmax(Iterator begin, Iterator end,
```

Comment:
It's useful to document that this is a vectorized version. i.e. it works with vectors of `NITEMS` floats.